### DISCUSSION TITLE
Can Firenvim be used to connect to a remote headless Neovim session?

### DISCUSSION THREAD

**Discussion by Kraust** [glacambre/firenvim#1634] [Q&A]
I am trying to do some research on a project I've wanted to work on for several years (Using the browser to connect to a headless neovim session) and I'm considering leveraging this either as a reference for how to render a nvim client in the browser, or maybe firenvim has already solved the problem for me.

## Use Case

I currently run Neovim Headless and connect to it over a reverse-SSH session. To do this I use Neovide as  It's the best performing Neovim client with good hardware acceleration (you an use nvim to do this now too). Figure below for reference:

```
[Server w/ Nvim] <---(TCP)--- [Client w/ Nvim]
```

I want to go a step further and do this:

```
[Server w/ Nvim] <---(TCP)--- [Client w/ Web Browser]
```

I try and explain this architecture to other Neovim users and they laugh at me because they don't understand why anyone would ever use a Browser to use Neovim. (For me it's Neovim everywhere). This project has similar design goals so I was hoping I could springboard a possible solution so I can start working on this as it has been bugging me for years now.

**Comment by glacambre** [comment]
Yes, on a technical level, it's possible. In fact, if you didn't care about security/authentication at all, you would only need to run neovim on your server and change [the server Firenvim connects to](https://github.com/glacambre/firenvim/blob/master/src/Neovim.ts#L26).

>For me it's Neovim everywhere

I have similar dreams :)

**Comment by Kraust** [comment]
I actually managed to pull this off with a bit of hacking but the performance isn't at the level I'm happy with. Doesn't look like firenvim scales well with a 4k textarea (But I wouldn't expect it to).

This was a really fun experiment tho. If anyone wants to do this in the future, this is roughly what I did:
- Changed the plugin to listen on a static port and token.
- Changed the client to connect to the static port (I still used localhost for this)
- Set up a reverse SSH Tunnel so that the remote neovim instance can be reached.
- Start up the remote neovim instance (running without --headless) and manually invoke start_server() with your token.
- IT works on small enough textareas. Performance goes out the door if you're doing things much beyond 1080p (I use so it won't work for me).

I had to stub out all of the qalls too as I only want to detach (but not quit) on detach. Overall if the performance was there (it might be, I might be missing something) Firenvim, could easily be adapted to work like Neovim QT & Neovide remote attach feature assuming you create a dummy web page with a textarea that spawns 100% of the page.

**Reply by glacambre** [reply]
Wow, very impressed to see that you figured how to do all this in a couple of hours and on your own!

Unfortunately you're not missing anything, Firenvim truly has terrible performance. There are three possible culprits that I can think of, from most likely to least likely:
1) The renderer itself. Firenvim is using the Canvas2D API, which is possibly the slowest API one can use to render arbitrary graphics. Moving to WebGL (or the newer WebGPU) might help.
2) Firenvim's websocket server. It's implemented in lua and I never even thought about optimizing its performance. Replacing it with an industrial-grade websocket-to-socket proxy could perhaps help.
3) The messagepack-rpc interface. At 4K, you probably have many lines/columns, so large grid updates could possibly create a lot of memory pressure. This is something I already had to optimize firenvim for in the past, but with the messagepack-rpc lib Firenvim uses, it might not be enough.

Perhaps you could try to compile Neovide to WASM? You'd likely only have to replace the parts that send/receive messages to/from neovim with something that opens a websocket, I'd be happy to advise you with that should you need help (although to be frank I doubt you would!).
